### PR TITLE
Direct communication link between user JS and JL with WebIO support

### PR DESCRIPTION
**Pull request by dralletje**
Some things are still needed before merging, like writing the small HTTP-interop types for making AssetRegistry more generic.

~Also, I just copied in Require to prevent a dependency, but I admit this is not the most elegant.~ EDIT: we decided that this is the simplest way to go

Also, tests

### CODE REVIEW

**Comment by fonsp** [comment]
wow!

**Comment by fonsp** [comment]
@dralletje Can you share your testing notebooks?

**Comment by dralletje** [comment]
Sii

**Comment by fonsp** [comment]
With an extra channel to trigger a bond update: https://github.com/fonsp/Pluto.jl/tree/webio-trigger-reactivity-api

**Comment by fonsp** [comment]
@SimonDanisch

**Comment by dralletje** [comment]
WebIO example: https://gist.github.com/dralletje/f390fbca683ccff05ebd6531e68eec73
Dash example: https://gist.github.com/dralletje/131eff9e90703346ca7db09ebcec223f

**Comment by dralletje** [comment]
Dash but even cooler (no actual server starting, all implicitly)
https://gist.github.com/dralletje/98187e573c5de39b65757cb6a3a72d5d

**Comment by fonsp** [comment]
One thing I realised is that AssetRegistry won't work with our cool new static state export. Right now, it's WYSIWYG, which is VERY cool, because everything goes through dralbase. 

Is there a way to make webio support work through dralbase? For example, can we put the assets in the statefile and use the serviceworker to capture the requests?

File: src/runner/IntegrationsWithOtherPackages.jl [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        )
+    end
+"""
+function on_request(module_name, request)::Dict{Symbol,<:Any}
+    throw(error("No http request handler defined for '$(module_name)'"))
+end
+
+"Still a bit unsure about this, but it will for now give you the (relative) path for your module requests"
+function get_base_url(module_name::Union{AbstractString, Symbol})
+    "/integrations/$(string(module_name))/$(workspace_info.notebook_id)"
```
Comment (fonsp) [inline src/runner/IntegrationsWithOtherPackages.jl:75]: Can we flip these two?

/integrations/$(workspace_info.notebook_id)/$(string(module_name))

File: src/runner/IntegrationsWithOtherPackages.jl [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        @error "Dispatching integrations websocket message failed:" message=message exception=(ex, bt)
+    end
+    nothing
+end
+const message_channel = Channel{Any}(10)
+
+"""
+Do not call yourself! Or do, but don't say I didn't warn you.
+Integrations should implement this to capture incoming websocket messages.
+I force returning nothing, because returning might give you the idea that
```
Comment (fonsp) [inline src/runner/IntegrationsWithOtherPackages.jl:33]: I think that the `::Nothing` on this method does not apply to the methods added by integrations, so we should typeassert somewhere else

**Comment by fonsp** [comment]
I was curious whether it is now also possible for packages to add integrations from within their package? e.g.

```julia
module MyPackage

function __init__()
	if isdefined(Main, :PlutoRunner)
		import Main.PlutoRunner.Integrations: on_ws_message, on_request
		
		...
	end
end
end
```

File: src/runner/IntegrationsWithOtherPackages.jl (hunk 2 above)
Comment (fonsp) [reply src/runner/IntegrationsWithOtherPackages.jl:33]: Sorry you already do that!

**Comment by fonsp** [comment]
How about those tests?

**Comment by fonsp** [comment]
this is very cool!!!

**Comment by fonsp** [comment]
TableView.jl works! @pfitzseb

https://gist.github.com/fonsp/9e4f5b0052e959427d9575cea049dd1d


https://user-images.githubusercontent.com/6933510/111298936-a3963280-864f-11eb-823b-06517b222c7e.mov

**Comment by pfitzseb** [comment]
Hey that's pretty awesome! :)

**Comment by pankgeorg** [comment]
WOOOOOOW

**Comment by dralletje** [comment]
For some reason PlotlyJS just stopped working 😱

**Comment by disberd** [comment]
Thanks a lot for the amazing work into trying to add WebIO support.

I have been trying this branch as I needed to access a trace type that is available in plotly.js (thus also via PlotlyJS.jl) but no through the Plots API ([sunburst charts](https://plotly.com/javascript/sunburst-charts/).
Everything works nicely in Pluto but plots can't be displayed in a static exported html as I get the WebIO not detected error as opposed to the plot output

image

I know this is still a WIP as it's not even yet merged in the master branch, but do you know whether this will be a problem that can be solved eventually or will the use of WebIO inside pluto break the stati html export functionality?

**Comment by fonsp** [comment]
@disberd WebIO won't work on statically generated HTML pages, because the goal of WebIO is to connect JavaScript to a live Julia process.

**Comment by disberd** [comment]
@fonsp Thanks for the clarification.
I see that indeed there was a similar issue when using PlotlyJS in jupiter [here](https://github.com/JuliaPlots/PlotlyJS.jl/issues/322).

I'll check to see if I'll be able to work around this by doing some static export like suggested at some point there, as I suppose in principle that is something that should be possible :).

**Comment by jbrea** [comment]
I am thinking about switching from Jupyter to Pluto for teaching this summer.
In Jupyter I use PlotlyJS.jl for plotting and I really like the interactivity (zooming, panning, etc.).
Any chance this PR - and with it PlotlyJS.jl in Pluto - will make it into a Pluto release before mid-July?

**Comment by disberd** [comment]
@jbrea If you are just interested in the basic pluto interactivity without need for fancy custom events handling (e.g. executing custom julia functions when clicking on plots)
You can already use the plotly library via PlotlyBase in pluto just by adding a simple cell of code do your notebooks.
I had a post on discourse on the matter here:
https://discourse.julialang.org/t/pluto-plots-with-plotlybase-plotlyjs-with-working-static-html-export/61730/12

This remove the need to have WebIO support and also enables working plots in static html export from Pluto.

Of course when this branch gets merged you will get the full WebIO interactivity as well using PlotlyJS, but the PlotlyBase functionality might be enough for your use case :)

**Comment by fonsp** [comment]
@jbrea Assume that it won't, that will be the best way to prepare for your semester!

**Comment by jbrea** [comment]
@disberd Cool workaround, thanks a lot!
@fonsp Thanks for the heads-up.

**Comment by sglyon** [comment]
This is awesome! 

If possible, I'd love to help get this over the line asap. I'm not intimately familiar with the Pluto codebase, but I could possibly help. 

What is holding this one up?

**Comment by SimonDanisch** [comment]
I wonder who will maintain WebIO though: https://github.com/JuliaGizmos/WebIO.jl/pull/460 ...

If I find some time, I could try to write a guide about how to switch from WebIO -> JSServe, which I'll definitely maintain as long as there is Makie...
@fonsp, @dralletje and I have been working at a similar integration for JSServe as part of this PR, so that could all work, I guess...

**Comment by sglyon** [comment]
I didn't realize that WebIO was so ... dead :(

Perhaps it would be best for me to move PlotlyJS to something else.

I haven't heard of JSServe.jl. Is that something that we think live on?

**Comment by SimonDanisch** [comment]
https://github.com/SimonDanisch/JSServe.jl
Well, as I said, I / we maintain it as part of Makie...Since Makie is doing pretty well right now, it should give it a pretty solid future ;)

In short, I developed it, because it was pretty impossible to write WGLMakie in a modular way based on WebIO - and things like offline support etc were also pretty impossible, and I never really liked the dom, observable + scope API, which I replaced by building upon an existing DOM abstraction, namely Hyperscript. 
It was also supposed to be more lightweight, but I guess at this point it rather offers a super set of WebIO 😅 

I've never really pushed it as an WebIO alternative, but now that WebIO is in a coma, it seems like a reasonable way forward...
The only problem right now is Jupyter support, which has been bit rotting away, and because of https://github.com/JuliaLang/IJulia.jl/issues/1002, I haven't had a chance to fix it yet...

**Comment by empet** [comment]
The above [issue](https://github.com/JuliaLang/IJulia.jl/issues/1002)  related to IJulia with Julia 1.16+ is not real. IJulia works on both  Windows 10 and Ubuntu 18.04

**Comment by SimonDanisch** [comment]
>The above issue related to IJulia with Julia 1.16+ is not real. I

lol, tell that both my Windows computers...

**Comment by pfitzseb** [comment]
I'd love JSServe as a WebIO alternative, FWIW. :)

**Comment by empet** [comment]
On my own it works :)

**Comment by dralletje** [comment]
@SimonDanisch you happen to remember what we came up with? I vague recall JSServe not having the same kind of hook-able-ness that WebIO provides, but that might be better now? I'd still love to ship JSServe compatibility with this PR

**Comment by SimonDanisch** [comment]
> I vague recall JSServe not having the same kind of hook-able-ness that WebIO provides, but that might be better now?

Not sure, i thought it was the other way around :D @fonsp and I took another look some time ago, and it looked like we could get things ready with a few lines of code ^^ But it's all a bit fuzzy

**Comment by travigd** [comment]
Hewwo, WebIO isn't dead, but could do with some more maintainers besides just me.

At the very least, I've updated WebIO for compat with Observables 0.4.

**Comment by diegozea** [comment]
Hi! Nice to hear that WebIO isn't dead :) What is the state of this PR? Best,

**Comment by sashmit** [comment]
> TableView.jl works! @pfitzseb @pankgeorg
> 
> https://gist.github.com/fonsp/9e4f5b0052e959427d9575cea049dd1d
> 
>  Schermopname.2021-03-16.om.12.02.52.mov

This tableview example (and indeed a lot of webio packages) don't seem to be working still.

**Comment by dralletje** [comment]
@sashmit Are you testing against this PR? This indeed isn't in Pluto still.

**Comment by sashmit** [comment]
> @sashmit Are you testing against this PR? This indeed isn't in Pluto still.

doh! I tried the webio branch, and it worked there. FYI, the version of webio can be bumped up one reevision here:
https://github.com/fonsp/Pluto.jl/pull/991/files#diff-7b616cf57a0d73b7dd6d3357b38ec19ded099ea2ab755bee15c006f3087c9c8dR28

**Comment by fonsp** [comment]
Superseded by https://github.com/fonsp/Pluto.jl/pull/2392

**Comment by fonsp** [comment]
Replaced by https://github.com/fonsp/Pluto.jl/pull/2726

**Pull request closed without merging**